### PR TITLE
feat: proposition power proposal validation strategy

### LINKED ISSUE #473: Whitelist proposal validation strategy
We need to add a whitelist proposal validation strategy

### LINKED ISSUE #494: Use `!=` instead of `==`
https://github.com/snapshot-labs/sx-starknet/blob/f9e623c49ea1a1969965eae1646012e9c36b01f7/starknet/src/types/indexed_strategy.cairo#L32

should be `!=`

### PR DESCRIPTION
**Pull request by pscott**
Adds the starknet version of [proposition power proposal validation strategy](https://github.com/snapshot-labs/sx-evm/blob/main/src/proposal-validation-strategies/PropositionPowerProposalValidationStrategy.sol)

replaces #476 

Closes #494 
Closes #473

### CODE REVIEW

**Comment by pscott** [comment]
btw we need to track https://github.com/software-mansion/scarb/discussions/568#discussioncomment-6742412  this way we can remove the `erc20_votes.cairo` file in the `mocks` folder

File: starknet/src/proposal_validation_strategies/proposing_power.cairo [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    use sx::utils::bits::BitSetter;
+    use box::BoxTrait;
+    use clone::Clone;
+
+    #[storage]
+    struct Storage {}
+
+    #[external(v0)]
+    impl ProposingPowerProposalValidationStrategy of IProposalValidationStrategy<ContractState> {
+        fn validate(
```
Comment (Orland0x) [inline starknet/src/proposal_validation_strategies/proposing_power.cairo:23]: probably makes sense to follow the same approach as sx-evm where we have the internal `_validate` function defined in `utils` then the strategy can call decode and call whichever ones it wants.

File: starknet/src/types/indexed_strategy.cairo [hunk 2]
```diff
@@ -29,7 +29,7 @@ impl IndexedStrategyImpl of IndexedStrategyTrait {
             // Check that bit at index `strats[i].index` is not set.
             let s = math::pow(u256 { low: 2_u128, high: 0_u128 }, *self.at(i).index);
 
-            assert((bit_map & s) == u256 { low: 1_u128, high: 0_u128 }, 'Duplicate Found');
+            assert((bit_map & s) != u256 { low: 1_u128, high: 0_u128 }, 'Duplicate Found');
```
Comment (Orland0x) [inline starknet/src/types/indexed_strategy.cairo:32]: nice catch

File: starknet/src/tests/proposal_validation_strategies/proposing_power.cairo [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            ProposingPowerProposalValidationStrategy::TEST_CLASS_HASH.try_into().unwrap(),
+            0,
+            array![].span(),
+            false
+        )
+            .unwrap();
+
+        let allowed_strategies = array![vanilla_strategy.clone()];
+        let proposal_threshold = 1_u256;
+        let mut params: Array<felt252> = array![];
```
Comment (Orland0x) [inline starknet/src/tests/proposal_validation_strategies/proposing_power.cairo]: cant infer the type here?

Comment (pscott) [reply starknet/src/tests/proposal_validation_strategies/proposing_power.cairo]: 8d28ee54bc422a0e512b6fba506907962cf8976c

File: starknet/src/proposal_validation_strategies/proposing_power.cairo (hunk 1 above)
Comment (pscott) [reply starknet/src/proposal_validation_strategies/proposing_power.cairo:23]: 0e70874af4f331fea537b7d4ea7f6566d3285da4

**Review by Orland0x — APPROVED** [review]
uTACK

**Pull request merged**